### DISCUSSION TITLE
Making it possible to search which files previously matched a specific search result

### DISCUSSION THREAD

**Discussion by CSyde65** [community/community#118891] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Product Feedback

### Body

There really needs to be a way to search which files previously matched a specific search result in addition to which files currently match a specific search result. Since I'm trying to search for a specific title across various files in a given repository. But I'm trying to search for which files previously matched a specific search result. Not just the files that currently match a specific search result.

I have tried to search for the following things in the [Wikia](https://github.com/Wikia) and [svn2github](https://github.com/svn2github) repositories:

`adminmentor` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from September 2012 at the latest.
`autoconfirmed` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be between 2004 and 2005.
`bot-global` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from August 2011 at the latest.
`chatmoderator` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be between May and July 2011.
`checkuser` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from March 2005 at the latest.
`checkuser-global` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from November 2010 at the latest.
`council` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from March 2011 at the latest.
`founder` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be between 2004 and 2005.
`helper` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from November 2007 at the latest.
`oversight` = the information I'm after may be located in the [Wikia](https://github.com/Wikia) repository, because it is estimated to be between July and August 2017.
`restricted-login` = the information I'm after may be located in the [Wikia](https://github.com/Wikia) repository, because it is estimated to be from October 2015 at the latest.
`rollback` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from May 2007 at the latest.
`staff` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from January 2006 at the latest.
`util` = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from August 2011 at the latest.
`wikiastars` = the information I'm after may be located in the [Wikia](https://github.com/Wikia) repository, because it is estimated to be between January and February 2013.
`wikifactory` = the information I'm after may be located in the [Wikia](https://github.com/Wikia) repository, because it is estimated to be from October 2015 at the latest.

The problem is that the revisions that contain the information that I'm after are past revisions, not current revisions. So searching for the things that I've already mentioned isn't helpful, because the only revisions that are listed are the latest revisions. And they don't contain the information that I'm looking for.

**Comment by LiteBrite82** [comment]
Hi @CSyde65!

Making sure I understand the question. I understand the gits of the question to be: You are trying to search the revision history of certain files? (and the current version)

**Reply by LiteBrite82** [reply]
so for example "adminmentor = the information I'm after may be located in the [svn2github](https://github.com/svn2github) repository, because it is estimated to be from September 2012 at the latest." 

You want to search the 1.3k repos in svn2github's profile for the term "adminmentor" and be able to see not only its current implementation in the Main branches, but also in the commit history and previous iterations.

**Reply by LiteBrite82** [reply]
(great examples by the way; well formed question) 🦸

**Reply by CSyde65** [reply]
Yes, that's right. I want to be able to see what files previously had 'adminmentor' in them, as opposed to what files currently have it in them.

**Reply by LiteBrite82** [reply]
You can search the [commit message history for the default branch](https://docs.github.com/en/search-github/searching-on-github/searching-commits).

Though thats only effective for devs that practice extreme detail. 

Other than that, you would have to go into the repo manually.

**Reply by LiteBrite82** [reply]
That being said...

This community is managed by GitHub employees, but it is not a support forum and there is no SLA or guarantee each post will get a response from a  product leader.

To get more traction with feature requests, checkout "Making Suggestions" [guidelines](https://github.com/community/community#making-suggestions) in the repo readme including: "Once you kick off a discussion, the GitHub product team will evaluate the feedback but will not be able to respond to every submission."

**Comment by CSyde65** [comment]
I have tried searching the commit history manually in the past. But the repositories are very large. So I can't guarantee that I'll be able to get the information I need in a timely manner. Which is why I wanted something more automatic. Something that can tell me what files have previously matched a specific input.

What I'm trying to do is:

- Figure out what month and year the `adminmentor` usergroup was added to Fandom. Estimated month and year: September 2012 at the latest. [[1]](https://github.com/svn2github/wikia/commits/13fe307a3782f5dab4777e26f8fe5a5fe3a07b88/extensions/wikia/UserProfilePageV3/strategies/UserTwoTagsStrategy.class.php)
- Figure out what month and year the `autoconfirmed` usergroup was added to Fandom. Estimated month and year: Between 2004 and 2005.
- Figure out what month and year the `bot-global` usergroup was added to Fandom. Estimated month and year: August 2011 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2011-08-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `chatmoderator` usergroup was added to Fandom. Estimated month and year: Between May and July 2011. [[1]](https://github.com/svn2github/wikia/commit/5dd4ebb9ed30e3b6c8520ea5c137a21a8626d3ee)[[2]](https://github.com/svn2github/wikia/commit/ebb598bb9fea27e8096bdf85d779c5287355490e)[[3]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2011-07-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `checkuser` usergroup was added to Fandom. Estimated month and year: March 2005 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2005-03-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `checkuser-global` usergroup was added to Fandom. Estimated month and year: November 2010 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2010-11-30&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `council` usergroup was added to Fandom. Estimated month and year: March 2011 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2011-03-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `founder` usergroup was added to Fandom. Estimated month and year: Between 2004 and 2005.
- Figure out what month and year the `helper` usergroup was added to Fandom. Estimated month and year: November 2007 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2007-11-30&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `oversight` usergroup was removed from Fandom. Estimated month and year: Between July and August 2017. [[1]](https://github.com/search?q=org%3AWikia+oversight&type=commits&s=committer-date&o=desc&p=1)
- Figure out what month and year the `restricted-login` usergroup was added to Fandom. Estimated month and year: October 2015 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2015-10-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `rollback` usergroup was added to Fandom. Estimated month and year: May 2007 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2007-05-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `staff` usergroup was added to Fandom. Estimated month and year: January 2006 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2006-01-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `util` usergroup was added to Fandom. Estimated month and year: August 2011 at the latest. [[1]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2011-08-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `wikiastars` usergroup was added to Fandom. Estimated month and year: Between January and February 2013. [[1]](https://github.com/Wikia/OldStructuredDataPrototype/commits/95f01633dc6305edc6655d20e36618985333279e/extensions/wikia/UserProfilePageV3/UserProfilePage.i18n.php)[[2]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2013-02-28&tagfilter=&subtype=&wpFormIdentifier=logeventslist)
- Figure out what month and year the `wikifactory` usergroup was added to Fandom. Estimated month and year: October 2015. [[1]](https://github.com/search?q=repo%3AWikia%2Fapp+wikifactory&type=commits&s=committer-date&o=asc&p=24)[[2]](https://community.fandom.com/wiki/Special:Log?type=rights&user=&page=&wpdate=2015-10-31&tagfilter=&subtype=&wpFormIdentifier=logeventslist)

**Reply by LiteBrite82** [reply]
do you work at Fandom? have access to their project boards (if they use them)

**Reply by LiteBrite82** [reply]
I was trying to think of another approach, like looking at the repo insights; I didn't see anything that stuck out there to me. I was hoping that they had an open project board to look at - the timeline feature in there would be perfect for your use case; though sadly, if they have one, its set to private....

**Reply by LiteBrite82** [reply]
I was able to track down 'adminmentor' to October 6th 2012 fairly quickly, just looking at the time stamps on the commit history of the files that contained `adminmentor`.
https://github.com/svn2github/wikia/commit/e93e593729489e5d1968f4cdda677fba15c616a8

I did have to do some manual searching; though minimally. 
I did a little cheat in the commit history by using ctrl + F to have the browser search while I was opening the commits.

**Reply by LiteBrite82** [reply]
I noticed the devs do have good commit messages, I've never searched for terms in commit messages myself, but it looks like we have that functionality. searching for any commit with `adminmentor` in the message may be a better way to quickly get to the date....

**Reply by LiteBrite82** [reply]
I just noticed the change log links you included for fandom; is that close to the format you want for the returns? Because that would be awesome!

**Reply by LiteBrite82** [reply]
I asked GitHub Copilot "How can I search for the first time a function was used in the commit history"
<img width="1678" alt="Screenshot 2024-04-17 at 1 09 17 PM" src="https://github.com/community/community/assets/163985914/9ea3f7c3-a006-4c4b-9dd2-f4643a362728">

And got an interesting answer:
how can i search for the first time a function was used in the commit history

Copilot:
Searching for the first time a function was used in the commit history can be done manually by using git log command in combination with grep. Here is a basic example:

Clone the repository to your local machine if you haven't already. You can do this with git clone <repository-url>.
Navigate to the repository directory with cd <repository-directory>.
Use the git log -S 'function_name' command. Replace 'function_name' with the name of the function you're looking for. This command will give you a list of commits where the usage of the function was added or removed.
To find the first instance, you may need to sift through the results from oldest to newest.
Please note that this approach may not be 100% accurate as it heavily depends on the commit messages and the changes in the commits. If the function was introduced without a proper commit message or in a commit that includes other changes, it can be missed.

Also, this method does not work directly on GitHub.com and needs to be performed on your local machine. As a GitHub Copilot Chat Assistant, I do not have the capability to perform this operation for you.

**Reply by LiteBrite82** [reply]
To speed up the process you could use GitHub CodeSpaces to run the repo virtually and avoid download times and clutter on your machine. 

I had forgotten how awesome `git log` is. I use it all the time to tell if students are cheating (10 commits in the span of an hour to complete an application that takes an average of 4 hours to complete)

**Comment by CSyde65** [comment]
They are links similar to the ones that I've been using to verify what month and year some of the other usergroups were introduced and/or retired. Although many of them were ones that I already had some idea of when they were introduced/retired. I haven't had it as easy with the ones that I've mentioned. Even with change logs, commits, and pull requests existing. Since I can't always tell by the change logs which revisions will contain what I'm after.

It's possible that some of the information that I'm after may have been deleted from the repositories. Which is why I wanted to be able to use a feature that would be able to tell me whether or not the information still exists. And if it still exists, which files previously contained the information that I'm after. I also linked various points of the Community Central user rights log, as I've been using that to get some idea when some of the usergroups were introduced/retired at the latest.

**Comment by CSyde65** [comment]
I don't have copilot enabled. So unless there's some way I can enable it for free, I'm not going to be able to find the information I'm after, myself.

**Reply by LiteBrite82** [reply]
CodeSpaces: free limited use
Copilot: are you a student or part of a larger org that has an enterprise account?

**Comment by CSyde65** [comment]
No I'm not a student, nor am I part of a larger organisation that has an enterprise account.

**Comment by LiteBrite82** [comment]
Something similar came up in [another discussion](https://github.com/orgs/community/discussions/56097#discussioncomment-9205422); its not exactly what we have been working on, but I wonder if its helpful.

**Comment by CSyde65** [comment]
Would it be possible to provide a step by step list of instructions of what I need to do in order to extract all the information that I'm looking for? I tried to get the free trial of Copilot, but it was too complicated for me to set up. As it required information that I don't know off the top of my head.

**Reply by LiteBrite82** [reply]
@look any thoughts?

**Reply by look** [reply, marked as answer]
At this time, GitHub does not maintain a search index that would make these queries possible. You will need to clone the repositories and use git commands to find what you're looking for. For example `git log -p -S`.

**Comment by CSyde65** [comment]
> At this time, GitHub does not maintain a search index that would make these queries possible. You will need to clone the repositories and use git commands to find what you're looking for. For example `git log -p -S`.

How do I do that?

**Reply by LiteBrite82** [reply]
Sorry to keep you hanging so long; busy weekend. 
The easiest web friendly way to do it is to open the repo in [CodeSpaces](https://www.youtube.com/shorts/rB9v6HoDXYo?app=desktop). Then, [run the `git log` commands in the terminal](https://www.youtube.com/watch?v=XLLqqucneh0).

GitHub Codespaces is free up to 120 "core hours" per month. Codespaces runs in a docker container that can be configured to use different hardware depending on your needs. In the case of "core hours" its referring to how many processor cores you are using and for how long. You can select anywhere from 2-32 cores; the free 120 hours will be divided by how many cores you select. 120 core hours / 2 cores = 60 hours of actual usage. [You can add more hours with a fee](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces).  

You can also do this on your local machine if you have VScode installed - though you have to clone the repo you want to search; which, depending on the repo, may or may not be practical. (which is why I suggested Codespaces) 

(Thanks for the git log tutorial @balt1794)

**Reply by LiteBrite82** [reply]
I found a bunch of great git log tutorials on you tube; [this was one of my favorites](https://www.youtube.com/watch?v=UZN7tPC4dHk) by @kardelio

**Reply by LiteBrite82** [reply]
Thanks @look

**Reply by Kardelio** [reply]
Thank you so much for the shout out @LiteBrite82 😊❤️😍 really appreciate it and glad you liked the video :)

**Reply by balt1794** [reply]
Thanks for the shout out @LiteBrite82 !
